cmd/outline-ss-server/config.go

```diff
@@ -29,15 +38,139 @@ type ServiceConfig struct {
+	TCPListenerType             = ListenerType("tcp")
+	UDPListenerType             = ListenerType("udp")
+	WebsocketStreamListenerType = ListenerType("websocket-stream")
+	WebsocketPacketListenerType = ListenerType("websocket-packet")
+)
+
+type WebServerConfig struct {
+	// Unique identifier of the web server to be referenced in Websocket connections.
+	ID string
 
-const listenerTypeUDP ListenerType = "udp"
+	// List of listener addresses (e.g., ":8080", "localhost:8081"). Should be localhost for HTTP.
+	Listeners []string `yaml:"listen"`
+}
 
+// ListenerConfig holds the configuration for a listener.  It supports different
+// listener types, configured via the embedded type and unmarshalled based on
+// the "type" field in the YAML/JSON configuration. Only one of the fields will
+// be set, corresponding to the listener type.
 type ListenerConfig struct {
-	Type    ListenerType
+	// TCP configuration for the listener.
+	TCP *TCPUDPConfig
+	// UDP configuration for the listener.
+	UDP *TCPUDPConfig
+	// Websocket stream configuration for the listener.
+	WebsocketStream *WebsocketConfig
+	// Websocket packet configuration for the listener.
+	WebsocketPacket *WebsocketConfig
+}
+
+var _ Validator = (*ListenerConfig)(nil)
+var _ yaml.Unmarshaler = (*ListenerConfig)(nil)
+
+// Define a map to associate listener types with [ListenerConfig] field names.
+var listenerTypeMap = map[ListenerType]string{
+	TCPListenerType:             "TCP",
+	UDPListenerType:             "UDP",
+	WebsocketStreamListenerType: "WebsocketStream",
+	WebsocketPacketListenerType: "WebsocketPacket",
+}
+
+func (c *ListenerConfig) UnmarshalYAML(value *yaml.Node) error {
+	var raw map[string]interface{}
+	if err := value.Decode(&raw); err != nil {
+		return err
+	}
+
+	// Remove the "type" field so we can decode directly into the target struct.
+	rawType, ok := raw["type"]
+	if !ok {
+		return errors.New("`type` field required")
+	}
+	delete(raw, "type")
+
+	lnType := ListenerType(rawType.(string))
+	fieldName, ok := listenerTypeMap[lnType]
+	if !ok {
+		return fmt.Errorf("invalid listener type: %v", lnType)
+	}
+	v := reflect.ValueOf(c).Elem()
+	field := v.FieldByName(fieldName)
+	if !field.IsValid() {
+		return fmt.Errorf("invalid field name: %s for type: %s", fieldName, lnType)
+	}
+	fieldType := field.Type()
+	if fieldType.Kind() != reflect.Ptr || fieldType.Elem().Kind() != reflect.Struct {
+		return fmt.Errorf("field %s is not a pointer to a struct", fieldName)
+	}
+
+	configValue := reflect.New(fieldType.Elem())
+	field.Set(configValue)
+	if err := mapstructure.Decode(raw, configValue.Interface()); err != nil {
+		return fmt.Errorf("failed to decode map: %w", err)
+	}
+	return nil
+}
+
+func (c *ListenerConfig) validate() error {
+	v := reflect.ValueOf(c).Elem()
+
+	for i := 0; i < v.NumField(); i++ {
+		field := v.Field(i)
+		if field.Kind() == reflect.Ptr && field.IsNil() {
+			continue
+		}
+		if field.Type().Implements(reflect.TypeOf((*Validator)(nil)).Elem()) {
```

Comment:
This is clearer (needs testing):
```suggestion
		if field.Type().Implements(reflect.TypeFor[Validator]()) {
```

But see below